ontogpt-habitat/main.py

```diff
@@ -0,0 +1,185 @@
+import logging
+import os
+import uuid
+import requests
+import datetime
+from typing import Tuple, Any, Dict, List
+#import sys
+#sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
+
+from shared import shared
+from kafka import KafkaConsumer, KafkaProducer
+
+logging.basicConfig(format="%(asctime)s - %(message)s", level=logging.INFO)
+
+
+def start_kafka() -> None:
+    """
+    Start a kafka listener and process the messages by unpacking the image.
+    When done it will republish the object, so it can be validated and stored by the processing service.
+    :param predictor: The predictor which will be used to extract the ontologies for dwc:habitat and dwc:locality digital specimen  
+
+    """
+    consumer = KafkaConsumer(
+        os.environ.get("KAFKA_CONSUMER_TOPIC"),
+        group_id=os.environ.get("KAFKA_CONSUMER_GROUP"),
+        bootstrap_servers=[os.environ.get("KAFKA_CONSUMER_HOST")],
+        value_deserializer=lambda m: json.loads(m.decode("utf-8")),
+        enable_auto_commit=True,
+    )
+
+    producer = KafkaProducer(
+        bootstrap_servers=[os.environ.get("KAFKA_PRODUCER_HOST")],
+        value_serializer=lambda m: json.dumps(m).encode("utf-8"),
+    )
+
+    for msg in consumer:
+        logging.info(f"Received message: {str(msg.value)}")
+        json_value = msg.value
+        try:
+            shared.mark_job_as_running(job_id=json_value.get("jobId"))
+            digital_object = json_value.get("object")
+            habitat = digital_object.get("ods:hasEvents")[0].get("dwc:habitat")
+            locality = digital_object.get("ods:hasEvents")[0].get("ods:hasLocation").get("dwc:locality")
+            additional_info_annotations = (
+                run_ontology_extraction(habitat, locality)
+            )
+            annotations = map_result_to_annotation(
+                digital_object, additional_info_annotations
+            )
+            annotation_event = map_to_annotation_event(annotations, json_value["jobId"])
+
+            logging.info(f"Publishing annotation event: {json.dumps(annotation_event)}")
+            publish_annotation_event(annotation_event, producer)
+
+        except Exception as e:
+            logging.error(f"Failed to publish annotation event: {e}")
+            send_failed_message(json_value["jobId"], str(e), producer)
+
+
+def map_to_annotation_event(annotations: List[Dict], job_id: str) -> Dict:
+    return {"annotations": annotations, "jobId": job_id}
+
+
+def publish_annotation_event(
+        annotation_event: Dict[str, Any], producer: KafkaProducer
+) -> None:
+    """
+    Send the annotation to the Kafka topic.
+    :param annotation_event: The formatted list of annotations
+    :param producer: The initiated Kafka producer
+    :return: Will not return anything
+    """
+    logging.info(f"Publishing annotation: {str(annotation_event)}")
+    producer.send(os.environ.get("KAFKA_PRODUCER_TOPIC"), annotation_event)
+
+
+def map_result_to_annotation(
```

Comment:
If ontoGPT didn't return a result, it would be nice to have that captured as an annotation. see `shared.send_empty_annotation`.